nixos/modules/services/web-servers/keter/default.nix

```diff
@@ -0,0 +1,105 @@
+      type = lib.types.package;
+      default = null;
+      description = ''set the keter package to be used, usually pkgs.haskellPackages.keter'';
+    };
+
+    globalKeterConfig = lib.mkOption {
+      type = lib.types.attrs;
+      default = {
+        ip-from-header = true;
+        listeners = [{
+            host = "*4";
+            port = 81;
+        }];
+      };
+      description = ''root is required to function in nixos .
+                      port 81 is set to default to allow nginx to reverse proxy into keter,
+                      so that nginx can do ssl management.
+                      You want that ip-from-header in the nginx setup case
+                      so it's not set to 127.0.0.1.
+                      Keep in mind root is overwritten by keterRoot.
+                      which is required for the nix derivation.
+                      '';
+    };
+
+    bundle = {
+        executable = lib.mkOption {
+            type = lib.types.str;
+            default = "bash";
+            description = ''The executable to be run in the bin folder of the package'';
+        };
+
+        domain = lib.mkOption {
+            type = lib.types.str;
+            default = "example.com";
+            description = ''The domain keter will bind to'';
+        };
+
+        # these won't work with standard nix options because they have
+        # to be put in a tarball, so we use a runtime script to load settings.
+        publicScript = lib.mkOption {
+            type = lib.types.str;
+            default = "";
+            description = ''
+                Allows loading of public environment variables,
+                these are emitted to the log so it shouldn't contain secrets.
+                For example:
+                ```
+                ADMIN_EMAIL=hi@example.com
+                ```
+            '';
+        };
+
+        secretScript = lib.mkOption {
+            type = lib.types.str;
+            default = "";
+            description = ''
+                Allows loading of private environment variables,
+                for example:
+                ```
+                MY_AWS_KEY=$(cat /run/keys/AWS_ACCESS_KEY_ID)
+                ```
+                '';
+        };
+
+        # This indirection is required to ensure the nix path
+        # gets copied over to the target machine in remote deployments.
+        # Furthermore, it's important that we use exec to
+        # run the binary otherwise we get process leakage.
+        package = lib.mkOption {
+            type = lib.types.package;
+            default = null;
+            description = ''The package which will be copied over to the nixos configuration,
+                            from which executable name runs.
+                        '';
+        };
+    };
+
+  };
+
+  config = lib.mkIf cfg.enable {
+    systemd.services.keter = keterConfig.keter;
+    systemd.services.keter-load-bundle = keterConfig.load-keter-bundle;
```

Comment:
If you're reading `config.nix` it's unclear that it's two systemd services, please inline that here or move the `systemd.services.*` part to `config.nix` too

Reply:
I inlined the config file in here, this get's rid of a lot of inderection as well :+1:

Reply:
I was thinking GitHub broke for me, but it's not inlined? Maybe git human error? Forgot to push?

Reply:
hehe, yes, turns out airports aren't great for pushing so it didn't end up online even though I made the comment :sweat_smile: